File: processing/ClubFitness/WaveGenerator.java
import processing.core.PApplet;

public class WaveGenerator extends Generator {

	public enum Shape {
		TRIANGLE,
		SQUARE,
		SIN
	}

	Shape shape;

	public WaveGenerator(PApplet p, String name, Shape shape, float amp, int phase, int period) {
		super(p, name, amp, phase, period);
		this.shape = shape;
	}

	public float value(int millis) {
		if (shape == Shape.TRIANGLE) {
			int modtime = ((millis+phase) % period);
			if (modtime < (period / 2)) {
				return p.map(modtime, 0.0f, period/2.0f, 0.0f, amp);
			} else {
				return p.map(modtime-period/2, 0.0f, period/2, amp, 0.0f);
			}
		} else if (shape == Shape.SQUARE) {
			int modtime = ((millis+phase) % period);
			if (modtime < (period / 2)) {
				return amp;
			} else {
				return 0.0f;
			}
		} else if (shape == Shape.SIN) {
			return amp * (p.sin((float)(millis+phase) / period * 2 * p.PI)*0.5f+0.5f);
		}

		return 0;
	}

}

File: python/bellechasse/src/bellechasse/generators/wave_generator.py
from typing import Optional
import math
from enum import Enum

from .generator import Generator
from ..util.math import value_map


class WaveGenerator(Generator):

    class Shape(Enum):
        TRIANGLE = 1
        SQUARE = 2
        SIN = 3

    def __init__(
        self,
        *,
        name: Optional[str] = None,
        amp: float = 0.5,
        period: int = 1000,
        phase: int = 0,
        offset: float = 0.5,
        shape: Shape = Shape.SIN
    ):
        super().__init__(name=name, amp=amp, offset=offset, period=period, phase=phase)

        if not isinstance(shape, self.Shape):
            raise TypeError("Shape Enum not provided")
        self.shape = shape

    def value(self, millis: int) -> float:
        if self.shape == WaveGenerator.Shape.TRIANGLE:
            modtime = (millis + self.phase + self.period / 4) % self.period
            if modtime < (self.period / 2):
                return value_map(
                    modtime,
                    0.0,
                    self.period / 2.0,
                    self.offset - self.amp,
                    self.offset + self.amp,
                )
            else:
                return value_map(
                    modtime - self.period / 2,
                    0.0,
                    self.period / 2,
                    self.offset + self.amp,
                    self.offset - self.amp,
                )
        elif self.shape == WaveGenerator.Shape.SQUARE:
            modtime = (millis + self.phase) % self.period
            if modtime < (self.period / 2):
                return self.offset + self.amp
            else:
                return self.offset - self.amp
        elif self.shape == WaveGenerator.Shape.SIN:
            return (
                self.amp * math.sin((millis + self.phase) / self.period * 2 * math.pi)
                + self.offset
            )

        return 0
